source/common/formatter/substitution_formatter.h

```diff
@@ -574,6 +574,20 @@ class FilterStateFormatter : public FormatterProvider {
   const bool is_upstream_;
 };
 
+/**
+ * FormatterProvider for ProxyProtocolTlvs stored in FilterState from StreamInfo.
+ */
+class ProxyProtocolTlvsFormatter : public FormatterProvider {
+public:
+  // FormatterProvider
+  absl::optional<std::string> format(const Http::RequestHeaderMap&, const Http::ResponseHeaderMap&,
+                                     const Http::ResponseTrailerMap&, const StreamInfo::StreamInfo&,
+                                     absl::string_view) const override;
+  ProtobufWkt::Value formatValue(const Http::RequestHeaderMap&, const Http::ResponseHeaderMap&,
```

Comment:
Just curious: what's the usage of this function on `FomatterProvider`? What requirements does it have?

Reply:
The `FormatterProvider` [interface](https://github.com/10gen/envoy-serverless/blob/6ea671b2dcbdbb5ff8bd49114aff4f977d794c56/envoy/formatter/substitution_formatter.h#L63-L78) specifies that `formatValue` is used to extract a value from the provided headers/trailers/stream, preserving the value's type, and the implementations of it also perform any formatting required by a formatter.

Reply:
Thanks! Where is this used? What's the difference of use cases from `format()`. I wonder why Protobuf is needed.

Reply:
It looks like `formatValue()` is used when we want to preserve the type of the value, and `format()` is used when we don't need to preserve the type (either by specification or because there are multiple formatter providers in use and there isn't just 1 type) and it's just formatted directly as a string. ([code link](https://github.com/10gen/envoy-serverless/blob/6ea671b2dcbdbb5ff8bd49114aff4f977d794c56/source/common/formatter/substitution_formatter.cc#L221-L251))

Some implementations of `formatValue()` call `format()` directly, presumably because the formatter provider would never need the type to be preserved. ([example](https://github.com/10gen/envoy-serverless/blob/6ea671b2dcbdbb5ff8bd49114aff4f977d794c56/source/common/formatter/substitution_formatter.cc#L1570-L1576))

Also, it looks like even the strings returned by `format()` are cast to `ProtobufWkt::Value` as the return value of `StructFormatter::providersCallback` ([link](https://github.com/10gen/envoy-serverless/blob/6ea671b2dcbdbb5ff8bd49114aff4f977d794c56/source/common/formatter/substitution_formatter.cc#L221)), so `ProtobufWkt::Value` is the desired type for either function 1 level up. 

I'm not too familiar with protobuf, so it's very possible that I'm missing something here. Please let me know if there's anything else you're thinking / looking for here?